Cloud alerts (#90)
astronomer/docs: cloud/stable/04_customize-airflow/04_airflow-alerts.md

> Deployment-level---

Author: Done!

Unlike task-level alerts, Deployment-level alerts are sent by Astronomer and do _not_ require a separate SMTP configuration.

![Astronomer Deployment Email Settings](https://assets2.astronomer.io/main/docs/emails/astro_deployment_email.png)

## Deployment Alerts on Astronomer

Refer to the following chart for more information on each deployment-level alert you might receive from Astronomer, as well as follow-up steps for improving the health of your affected Deployments.

| Alert | Description | Follow-Up |
| ------------- | ------------- | ------------- |
| `AirflowDeploymentUnhealthy` | The Airflow Deployment is unhealthy or not completely available. | Contact your Astronomer representative. |
| `AirflowEphemeralStorageLimit` | The Airflow Deployment is nearing its limit for storing temporary data. | Ensure that you are continually removing unused temporary data in your Airflow tasks. |

> | `AirflowPodQuota` | The Airflow Deployment is near its pod quota; it's been using over 95% of it's pod quota for over 10 minutes. | Either increase the Deployment's allocated Webserver and Scheduler Resources or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |

Author: Assuming this was for the `AiflowPodQuota`. Should this be the only follow-up to recommend the Extra Capacity slider, rather than the allocated resources sliders?

> | `AirflowPodQuota` | The Airflow Deployment is near its pod quota; it's been using over 95% of it's pod quota for over 10 minutes. | Either increase the Deployment's allocated Webserver and Scheduler Resources or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |

Reviewer: @jwitz correct. the resource sliders are specific to cpu/mem on the specific related components (webserver, scheduler, workers). the only thing that affects the pod quota is the extra capacity slider.

| `AirflowSchedulerUnhealthy` | The Airflow scheduler's heartbeat has dropped below the acceptable rate. | Contact your Astronomer representative. |
| `AirflowTasksPendingIncreasing` | The Airflow Deployment is creating tasks faster than it's clearing them. | Ensure that your tasks are running and completing correctly. If your tasks are running as expected, increase the Deployment's allocated Server Resources in the Astronomer UI or redesign your DAGs so that less tasks are concurrently pending. |

> | `ContainerMemoryNearTheLimitInDeployment` | The Airflow Deployment is near its memory quota; it's been using over 95% of it's memory quota for over 10 minutes. | Either increase the Deployment's allocated Webserver Resources or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |

Author: Sounds good! Updated

  ```
## Subscribe to Deployment Alerts on Astronomer

In the Astronomer UI, you can subscribe to Deployment-level alerts by navigating to the **Settings** tab of your Airflow Deployment and entering your email address under **Email Alerts**.

Astronomer's platform monitors the health of your Airflow Deployment and triggers an alert if any of Airflow's core components are underperforming (e.g. Scheduler, Webserver) or if you've initiated a faulty action. For instance, you may receive an alert if your Airflow Scheduler is unhealthy, if tasks are failing at an abnormal rate, or if you've attempted to upgrade to a version of Airflow that does not match the corresponding Docker image in your Dockerfile.

Unlike task-level alerts, Deployment-level alerts are sent by Astronomer and do _not_ require a separate SMTP configuration.

![Astronomer Deployment Email Settings](https://assets2.astronomer.io/main/docs/emails/astro_deployment_email.png)

## Deployment Alerts on Astronomer

> Refer to the following chart for more information on each deployment-level alert you might receive from Astronomer, as well as follow-up steps for improving the health of your affected Deployments.

Reviewer 2: ```suggestion
Refer to the chart below for more information on each Deployment-level alert you might receive from Astronomer, as well as corresponding remediation steps we recommend for your affected Deployment.
```

Author: Made a different suggestion on this, since I realized this my original wording is grammatically inconsistent (the follow-up step _is_ more information, so we are not showing that "as well as" more information). Also tied to make it more active.

| Alert | Description | Follow-Up |
| ------------- | ------------- | ------------- |

> | `AirflowDeploymentUnhealthy` | The Airflow Deployment is unhealthy or not completely available. | Contact your Astronomer representative. |

Reviewer 2: ```suggestion
| `AirflowDeploymentUnhealthy` | The Airflow Deployment is unhealthy or not completely available. | Reach out to [Astronomer Support](https://support.astronomer.io). |
```

Astronomer Cloud users/customers don't have a "representative" since our CS team doesn't have enough dedicated resources to reach all of them directly (and they likely won't be able to respond in an actionable way to notice of these alerts either), so I think our best bet is to send these folks to our support team.

> | `AirflowEphemeralStorageLimit` | The Airflow Deployment is nearing its limit for storing temporary data. | Ensure that you are continually removing unused temporary data in your Airflow tasks. |

Reviewer 2: ```suggestion
| `AirflowEphemeralStorageLimit` | The Airflow Deployment is nearing its limit for storing temporary data. | Make sure to continually remove unused temporary data in your Airflow tasks. |
```

Or even: 

`Remove unused temporary data in your Airflow tasks.`

> | `AirflowPodQuota` | The Airflow Deployment is near its pod quota; it's been using over 95% of it's pod quota for over 10 minutes. | Either increase the Deployment's Extra Capacity in the Astronomer UI or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |

Reviewer 2: Typo on `over 95% of it's pod quota` --> `over 95% of its pod quota`

Author: bleh, fixed

> | `AirflowSchedulerUnhealthy` | The Airflow scheduler's heartbeat has dropped below the acceptable rate. | Contact your Astronomer representative. |

Reviewer 2: ```suggestion
| `AirflowSchedulerUnhealthy` | The Airflow scheduler's heartbeat has dropped below the acceptable rate. | Reach out to [Astronomer Support](https://support.astronomer.io). |
```

> | `AirflowSchedulerUnhealthy` | The Airflow scheduler's heartbeat has dropped below the acceptable rate. | Reach out to [Astronomer Support](https://support.astronomer.io). |

Reviewer 2: ```suggestion
| `AirflowSchedulerUnhealthy` | The Airflow Scheduler's heartbeat has dropped below the acceptable rate. | Reach out to [Astronomer Support](https://support.astronomer.io). |
```

I think we typically capitalize Scheduler + Webserver - maybe something to standardize? Do we know what the acceptable rate of the heartbeat is?

Author: Good to know, I'll add that to the style guide.

It looks to be "no heartbeat for more than one minute" based on this line in the Prometheus code: `rate(airflow_scheduler_heartbeat{type="counter"}[1m]) == 0`
I'll add that in!

> | `AirflowTasksPendingIncreasing` | The Airflow Deployment is creating tasks faster than it's clearing them. | Ensure that your tasks are running and completing correctly. If your tasks are running as expected, increase the Deployment's allocated Server Resources in the Astronomer UI or redesign your DAGs so that less tasks are concurrently pending. |

Reviewer 2: Yikes really not sure what the recommended action item is here, so double check with @wolfier or @krisdock here - these guidelines sound like they're more telling you what's happening and what "should" happen but not so much how to get there.

When we mention `Server resources` though - is it Webserver, Scheduler, or Worker AU's they have to increase?

Author: Whoops - I meant to write Webserver Resources there. And yeah, I agree we shouldn't move forward with this until Allan or Kris get eyes on it.

Reviewer 3: Server resources is most likely referring to the scheduler since the scheduler is what pushes the tasks through its lifecycle. However, since tasks are going into the scheduled/queued state that means the scheduler is functioning correctly most likely. The executor could be at fault for not sending the tasks to the workers.

The more likely scenario is that the worker does not have the ability to consume the tasks from the queue. Then the solution becomes increasing worker productivity* for the Executor, which are generally KubernetesExecutor or CeleryExecutor.

Increasing worker productivity is a subject for another day ;p 

Basically tuning the worker/scheduler would resolve this alert and I don't think it's as simple as bumping resources.

> | `AirflowTasksPendingIncreasing` | The Airflow Deployment is creating tasks faster than it's clearing them. | Ensure that your tasks are running and completing correctly. If your tasks are running as expected, increase the Deployment's allocated resources in the Astronomer UI or redesign your DAGs so that fewer tasks are concurrently pending. |

Reviewer 2: - Should this be "If your tasks _aren't_ running as expected, increase...."?
- What resources should they modify?
- Instead of "Redesign your DAGs" I'd say -  "Lower Airflow Concurrency or Parallelism so that fewer tasks are concurrently pending."

Reviewer 2: Could maybe even link out to: https://www.astronomer.io/guides/airflow-scaling-workers

Author: I think they should do these things if their tasks _are_ running as expected, eg. their DAGs are properly creating and running tasks, but there still isn't enough resources to handle the number of pending tasks they have. 

Based on the article you linked out, I assume that the allocated resource they need to update will vary based on how they decide to increase concurrency/parallelism. In that case, I thin kit makes sense to link out to the guide for more information.

> | `ContainerMemoryNearTheLimitInDeployment` | The Airflow Deployment is near its memory quota; it's been using over 95% of it's memory quota for over 10 minutes. | Either increase the Deployment's allocated Webserver Resources or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |

Reviewer 2: Do we know what the memory quota is? The Airflow Webserver just parses DAG files and powers the Airflow UI so I'm not immediately seeing a connection - maybe just double check? Is it Webserver, Scheduler or Worker resources that would help? Or some combination?

Reviewer 3: @jwitz, you should check the actual [prometheus alerts](https://prometheus.gcp0001.us-east4.astronomer.io/rules) to verify the definition. For example `ContainerMemoryNearTheLimitInDeployment` is the collection of all the pods with some exceptions.

```
 (container_memory_working_set_bytes{container=~".*(scheduler|scheduler-gc|webserver|worker|statsd|pgbouncer|metrics-exporter|redis|flower)",container_name!~"POD|istio-proxy"} / container_spec_memory_limit_bytes{container=~".*(scheduler|scheduler-gc|webserver|worker|statsd|pgbouncer|metrics-exporter|redis|flower)",container_name!~"POD|istio-proxy"}) * 100 > 95 < +Inf
```

I would ask someone on infra to help you understand what those metrics mean.

Author: Sounds good! Updated all of the inaccurate thresholds in the latest pull

> | `ContainerMemoryNearTheLimitInDeployment` | A container in your Aiirflow Deployment is near its memory quota; it's been using over 95% of it's memory quota for over 60 minutes. | Either increase the Deployment's allocated resources in the Astronomer UI or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |

Reviewer 2: typo on "it's" v" its"

Reviewer 2: (on `ContainerMemoryNearTheLimitInDeployment`)

Author: Fixed!
